pkg/ingester/limiter_test.go

```diff
@@ -449,14 +662,66 @@ func TestLimiter_AssertMaxSeriesPerUser(t *testing.T) {
 			}, nil)
 			require.NoError(t, err)
 
-			limiter := NewLimiter(limits, ring, testData.ringReplicationFactor, false, "")
+			strategy := newIngesterRingLimiterStrategy(ring, testData.ringReplicationFactor, false, "", limits.IngestionTenantShardSize)
+			limiter := NewLimiter(limits, strategy)
 			actual := limiter.IsWithinMaxSeriesPerUser("test", testData.series, limiter.getShardSize("test"))
 
 			assert.Equal(t, testData.expected, actual)
 		})
 	}
 }
 
+func TestLimiter_AssertMaxSeriesPerUser_WithPartitionsRing(t *testing.T) {
+	tests := map[string]struct {
+		maxGlobalSeriesPerUser int
+		partitionStates        []ring.PartitionState
+		series                 int
+		minLocalLimit          int
+		expected               bool
+	}{
+		"limit is disabled": {
+			maxGlobalSeriesPerUser: 0,
+			partitionStates:        []ring.PartitionState{ring.PartitionActive},
+			series:                 100,
+			expected:               true,
+		},
+		"current number of series is below the limit": {
+			maxGlobalSeriesPerUser: 900,
+			partitionStates:        []ring.PartitionState{ring.PartitionActive, ring.PartitionActive, ring.PartitionActive},
+			series:                 299,
+			expected:               true,
+		},
+		"current number of series is above the limit, min limit not set": {
+			maxGlobalSeriesPerUser: 900,
+			partitionStates:        []ring.PartitionState{ring.PartitionActive, ring.PartitionActive, ring.PartitionActive},
+			series:                 300,
+			expected:               false,
+		},
+		"current number of series is above the limit, but below min limit": {
+			maxGlobalSeriesPerUser: 1000,
+			partitionStates:        []ring.PartitionState{ring.PartitionActive, ring.PartitionActive, ring.PartitionActive},
+			series:                 300,
+			minLocalLimit:          400,
+			expected:               true,
+		},
+	}
+
+	for testName, testData := range tests {
+		t.Run(testName, func(t *testing.T) {
+			pr := buildPartitionRingFromPartitionStates(testData.partitionStates...)
+
+			limits, err := validation.NewOverrides(validation.Limits{MaxGlobalSeriesPerUser: testData.maxGlobalSeriesPerUser}, nil)
+			require.NoError(t, err)
+
+			strategy := newPartitionRingLimiterStrategy(&partitionRingHolder{pr: pr}, limits.IngestionTenantShardSize)
+			limiter := NewLimiter(limits, strategy)
+			actual := limiter.IsWithinMaxSeriesPerUser("test", testData.series, testData.minLocalLimit)
```

Comment:
The 3rd param in this PR is not min local limit but the shard size. I think the cases in this test need to be redone.

Reply:
Thank you, good catch. I've fixed this test and also added test cases to `TestLimiter_AssertMaxSeriesPerUser` to pass non-zero shard size.